common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -183,6 +192,22 @@ public static int findInSet(final UTF8String match, final UTF8String set, int co
     return 0;
   }
 
+  public static int lowercaseIndexOf(final UTF8String target, final UTF8String pattern,
+      final int start) {
+    if (pattern.numChars() == 0) return 0;
+    int lenHaystack = target.numChars(), lenNeedle = pattern.numChars();
+    final UTF8String firstLower = pattern.substring(0, 1).toLowerCase();
+    final UTF8String needle = pattern.toLowerCase();
+    for (int i = start; i <= (lenHaystack - lenNeedle); i++) {
+      if (target.substring(i, i + 1).toLowerCase().equals(firstLower)) {
```

Comment:
What happens if `target` contains `i̇` at positions i and i+1 and `firstLower` is `İ`? I think this would give the wrong result.

Consider for example: `l` equals `i̇i̇` and `r` is `İ`. We want to find, if exists, the first a position `p` and a length `n` for `l` such that `substr(l, p, n)` lowercase-equals `r`. Here the return value for the function should be 0 because for  `p=0`  and `n=4` we have a match.

Reply:
agreed, `firstLower` concept unfortunately doesn't hold here - removed in latest commit